File: 2015-heat-humidity/hh_processAsosPrecip.m
% this reads pre-processed state-station ASOS hourly wx files, processed by
% ag_processAsosPrecip.py

baseDir = 'e:/data/projects/heat/asos/wx-data/';
%baseDir = '2017-ag-precip/wx-data/';

states = {'in', 'india'};

% this is for comma delimited ASOS data with the format:
% year, month, day, hour, lon, lat, tempC, relH, surf pressure, mslp
% -999 indicates a missing value
fileFormatStr = '%n %n %n %n %n %n %n %n %n %n %*[^\n]';

% process all files
for s = 1:length(states)
    state = states{s};
    
    asosData = {};
    
    curDir = [baseDir state];
    txtFileNames = dir([curDir, '/*.txt']);
    txtFileNames = {txtFileNames.name};
    
    if exist([baseDir 'asos-' state '.mat'], 'file')
        continue;
    end
    
    for f = 1:length(txtFileNames)
        fileName = txtFileNames{f};
        
        % split the file name at the . and take the first part to get the
        % station code
        fileNameParts = strsplit(fileName, '.');
        code = fileNameParts{1};
        
        ['processing ' state '/' code '...']
        
        fin = fopen([curDir '/' fileName]);
    
        % read weather file
        [data, pos] = textscan(fin, fileFormatStr, 'Delimiter', ',');

        % close current file
        fclose(fin);
        
        % columns:
        % 1 - year
        % 2 - month
        % 3 - day
        % 4 - hour
        % 5 - lon
        % 6 - lat
        % 7 - temp (C, -999 = missing value)
        % 8 - rel humidity (%, -999 = missing value)
        % 9 - surface pressure (in)
        % 10 - mslp (mb)

        years = data{1};
        months = data{2};
        days = data{3};
        hours = data{4};
        
        lons = data{5};
        lats = data{6};
        
        temps = data{7};
        temps(temps == -999) = NaN;
        
        relHs = data{8};
        relHs(relHs == -999) = NaN;
        
        % convert in to Pa
        surf = data{9} ;
        surf(surf == -999) = NaN;
        surf = surf .* 3386.39;
        
        % convert mb to Pa
        mslp = data{10};
        mslp(mslp == -999) = NaN;
        mslp = mslp .* 133.322;
        
        wb = [];
        
        % now calculate wet bulb for all obs
        ['computing wet bulb...']
        for i = 1:length(temps)
            % if we have all data for this timestep...
            if ~isnan(temps(i)) && ~isnan(relHs(i)) && ~isnan(surf(i))
                wb(i) = kopp_wetBulb(temps(i), surf(i), relHs(i), 1);
            else
                wb(i) = NaN;
            end
        end
        
        asosData{end+1} = {code, lons(1), lats(1), years, months, days, hours, temps, relHs, surf, mslp, wb};
    end
    
    save([baseDir 'asos-' state '.mat'], 'asosData', '-v7.3');
end

File: 2015-heat-humidity/hh_processAsosPrecip.py
# divide the main file into subfiles, one for each station

import os

asosBase = 'e:/data/projects/heat/asos/'
states = ['in']

# how many lines to read at a time
N = 1000000

def writeOutput(stationData, outputDir):
    print('writing output files...')
    # loop over all stations
    for code in stationData.keys():
        print('writing', code, '...')
        if code == 'PRN' or code == 'CON':
            fout = open(outputDir + code + '-valid' + '.txt', 'a')    
        else:
            fout = open(outputDir + code + '.txt', 'a')

        for lineParts in stationData[code]:
            # if rel hum exists, delete it
            if len(lineParts) == 9:
                del lineParts[7]
            fout.write(','.join(lineParts) + '\n')
        
        fout.close();

for state in states:
    outputDir = asosBase + 'wx-data/' + state + '/'
    if not os.path.exists(outputDir):
        os.makedirs(outputDir)
    else:
        continue
    
    fin = open(asosBase + 'asos-' + state + '.txt', 'r')

    # skip first lines of headers
    startLine = 1

    # current line number
    lineCnt = 1
    
    # last time we wrote to output files
    lastLineReset = 0

    # dictionary with station codes (keys) and data (values)
    stationData = {}

    print('reading ' + state + ' input files...')
    for line in fin:

        # remove whitespace (and trailing \n)
        line = line.strip()
        
        # if we're past the headers
        if lineCnt > startLine:
            lineParts = line.split(',')

            # if we haven't seen the current station yet, set it to empty list
            if lineParts[0] not in stationData.keys():
                stationData[lineParts[0]] = []

            # columns:
            # 0 - station
            # 1 - date/time
            # 2 - lon
            # 3 - lat
            # 4 - temp (C)
            # 5 - rel humidity (%)
            # 6 - surface pressure (in)
            # 7 - mslp (mb)

            # split time into separate cols for year, month, day, hour
            datetimeParts = lineParts[1].split(' ')

            dateParts = datetimeParts[0].split('-')
            timeParts = datetimeParts[1].split(':')
            
            year = int(dateParts[0])
            month = int(dateParts[1])
            day = int(dateParts[2])
            hour = int(timeParts[0])

            datetime = [str(year), str(month), str(day), str(hour)]

            # replace 'M' with '-999' in cols 2, 3, 4, 5, 6
            for i in range(2, len(lineParts)):
                lineParts[i] = lineParts[i].replace('M', '-999')

            # add data to station
            stationData[lineParts[0]].append(datetime + lineParts[2:])
        
        lineCnt += 1
        if lineCnt % 100000 == 0:
            print('line', lineCnt)

        # write to output file if we have enough data
        if lineCnt - lastLineReset > N:
            writeOutput(stationData, outputDir)
            lastLineReset = lineCnt
            stationData = {}

    # write remaining data to output at the end
    writeOutput(stationData, outputDir)

    
